lib/puppet/type.rb

```diff
@@ -1073,7 +1096,8 @@ def retrieve
     # Provide the name, so we know we'll always refer to a real thing
     result[:name] = self[:name] unless self[:name] == title
 
-    if ensure_prop = property(:ensure) or (self.class.needs_ensure_retrieved and self.class.validattr?(:ensure) and ensure_prop = newattr(:ensure))
+    ensure_prop = property(:ensure) || (self.class.needs_ensure_retrieved && self.class.validattr?(:ensure) && newattr(:ensure))
```

Comment:
If `property(:ensure)` returns nil, then it's a little magical to me that the second condition returns the value of the last component `newattr(:ensure)`, which we assign to `ensure_prop`. My brain sees `(... && ... && ...)` and thinks it returns a boolean (but it doesn't), which isn't what we want.